app/controllers/flood_risk_engine/enrollments/addresses_controller.rb

```diff
@@ -0,0 +1,107 @@
+require_dependency "flood_risk_engine/application_controller"
+
+module FloodRiskEngine
+  module Enrollments
+    class AddressesController < ApplicationController
+
+      def new
+        build_address
+        check_for_errors
+      end
+
+      def create
+        build_address
+        update
+      end
+
+      def edit
+        check_for_errors
+      end
+
+      def update
+        clear_error_params
+        redirect_to target_url
+      end
+
+      private
+
+      def form
+        @form ||= AddressForm.new(enrollment, address)
+      end
+      helper_method :form
+
+      def target_url
+        if save_form!
+          step_forward
+          enrollment_step_path(enrollment, enrollment.current_step)
+        else
+          session[:error_params] = {
+            address: params[:address]
+          }
+          failure_url
+        end
+      end
+
+      def failure_url
+        if params[:action] == "create"
+          new_enrollment_address_path(
+            enrollment,
+            address_initial_attributes.merge(
+              check_for_error: true,
+              step_back_to_postcode: params[:step_back_to_postcode]
+            )
+          )
+        else
+          edit_enrollment_address_path(
+            enrollment,
+            address,
+            check_for_error: true
+          )
+        end
+      end
+
+      def save_form!
```

Comment:
I think inheriting from a base class will obscure the behaviour of the controller. 

At the moment, the controller functionality is visible within a single class. Adding a base class would mean that functionality would be defined in two places.

If this control becomes difficult to read or overly complicated, then a base class may be a good way to deal with those problems. I don't believe that is the case here.